src/Mod/Sketcher/App/planegcs/GCS.cpp

```diff
@@ -1326,13 +1326,63 @@ int System::addConstraintInternalAlignmentParabolaFocus(Parabola &e, Point &p1,
     return addConstraintEqual(e.focus1.y, p1.y, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
-int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, int poleindex, int tagId, bool driving)
+int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, unsigned int poleindex, int tagId, bool driving)
 {
     addConstraintEqual(b.poles[poleindex].x, c.center.x, tagId, driving, Constraint::Alignment::InternalAlignment);
     addConstraintEqual(b.poles[poleindex].y, c.center.y, tagId, driving, Constraint::Alignment::InternalAlignment);
     return addConstraintEqual(b.weights[poleindex], c.rad, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
+int System::addConstraintInternalAlignmentKnotPoint(BSpline &b, Point &p, unsigned int knotindex, int tagId, bool driving)
+{
+    if (b.periodic && knotindex == 0) {
+        // This is done here since knotpoints themselves aren't stored
+        addConstraintP2PCoincident(p, b.start, tagId, driving);
+        addConstraintP2PCoincident(p, b.end, tagId, driving);
+    }
+
+    size_t numpoles = b.degree - b.mult[knotindex] + 1;
+    if (numpoles == 0)
+        numpoles = 1;
+
+    size_t startpole = 0;
+    std::vector<double *> pvec;
+    pvec.push_back(p.x);
+
+    std::vector<double> weights(numpoles, 1.0 / numpoles);
+
+    // TODO: Adjust for periodic knot
+    for (size_t j = 1; j <= knotindex; ++j)
+        startpole += b.mult[j];
+    if (!b.periodic && startpole >= b.poles.size())
```

Comment:
Consider a single-piece cubic B-spline. Its knots would be `[0,1]` with mults `[4,4]`. It would have 4 poles, so the last pole would have index 0. But this algorithm will give 4. This will happen for all non-periodic B-splines.

Meanwhile, for a Bspline with knots `[0,0.5,1]` and mults `[4,1,4]` the `startpole` for the second knot (`knotindex=1`) would be 1. The sum works for all knots except the last.

The best way here would be to think in terms of "flattened knots", perhaps written in a comment.